File: test/lexer.test.ts
import Lexer from '../src/lexer'
import { ASSIGN, COMMA, CONST, EOF, FUNCTION, IDENT, INT, LBRACE, LET, LPAREN, PLUS, RBRACE, RPAREN, TokenType } from '../src/token'

describe('Lexer', () => {
  describe('readChar()', () => {
    it('should read the next character of an input', () => {
      const input = 'wouldyoumind'
      const l = new Lexer(input)

      expect(l.char).toBe('w')

      for (let i = 1; i < input.length; i++) {
        l.readChar()
        expect(l.char).toBe(input[i])
      }
    })
  })

  describe('nextToken', () => {
    it('should map literals to tokens', () => {
      const input = `
        wouldyoumind summing(x, y) {
          x + y
        }

        permit five = 5
        insist ten = 10
        insist result = summing(five, ten)
      `

      const tests: { expectedType: TokenType, expectedLiteral: string }[] = [
        { expectedType: FUNCTION, expectedLiteral: 'wouldyoumind' },
        { expectedType: IDENT, expectedLiteral: 'summing' },
        { expectedType: LPAREN, expectedLiteral: '(' },
        { expectedType: IDENT, expectedLiteral: 'x' },
        { expectedType: COMMA, expectedLiteral: ',' },
        { expectedType: IDENT, expectedLiteral: 'y' },
        { expectedType: RPAREN, expectedLiteral: ')' },
        { expectedType: LBRACE, expectedLiteral: '{' },
        { expectedType: IDENT, expectedLiteral: 'x' },
        { expectedType: PLUS, expectedLiteral: '+' },
        { expectedType: IDENT, expectedLiteral: 'y' },
        { expectedType: RBRACE, expectedLiteral: '}' },
        { expectedType: LET, expectedLiteral: 'permit' },
        { expectedType: IDENT, expectedLiteral: 'five' },
        { expectedType: ASSIGN, expectedLiteral: '=' },
        { expectedType: INT, expectedLiteral: '5' },
        { expectedType: CONST, expectedLiteral: 'insist' },
        { expectedType: IDENT, expectedLiteral: 'ten' },
        { expectedType: ASSIGN, expectedLiteral: '=' },
        { expectedType: INT, expectedLiteral: '10' },
        { expectedType: CONST, expectedLiteral: 'insist' },
        { expectedType: IDENT, expectedLiteral: 'result' },
        { expectedType: ASSIGN, expectedLiteral: '=' },
        { expectedType: IDENT, expectedLiteral: 'summing' },
        { expectedType: LPAREN, expectedLiteral: '(' },
        { expectedType: IDENT, expectedLiteral: 'five' },
        { expectedType: COMMA, expectedLiteral: ',' },
        { expectedType: IDENT, expectedLiteral: 'ten' },
        { expectedType: RPAREN, expectedLiteral: ')' },
        { expectedType: EOF, expectedLiteral: '' }
      ]

      const l = new Lexer(input)

      for (const test of tests) {
        const token = l.nextToken()
        expect(token.type).toBe(test.expectedType)
        expect(token.literal).toBe(test.expectedLiteral)
      }
    })
  })
})

File: lexer/lexer_test.go
package lexer

import (
	"testing"

	"github.com/vherus/politescript/token"
)

func TestNextToken(t *testing.T) {
	input := `wouldyoumind summing(x, y) {
		x + y
	}

	permit five = 5
	insist ten = 10
	insist result = summing(five, ten)
	`

	tests := []struct {
		expectedType    token.TokenType
		expectedLiteral string
	}{
		{expectedType: token.FUNCTION, expectedLiteral: "wouldyoumind"},
		{expectedType: token.IDENT, expectedLiteral: "summing"},
		{expectedType: token.LPAREN, expectedLiteral: "("},
		{expectedType: token.IDENT, expectedLiteral: "x"},
		{expectedType: token.COMMA, expectedLiteral: ","},
		{expectedType: token.IDENT, expectedLiteral: "y"},
		{expectedType: token.RPAREN, expectedLiteral: ")"},
		{expectedType: token.LBRACE, expectedLiteral: "{"},
		{expectedType: token.IDENT, expectedLiteral: "x"},
		{expectedType: token.PLUS, expectedLiteral: "+"},
		{expectedType: token.IDENT, expectedLiteral: "y"},
		{expectedType: token.RBRACE, expectedLiteral: "}"},
		{expectedType: token.LET, expectedLiteral: "permit"},
		{expectedType: token.IDENT, expectedLiteral: "five"},
		{expectedType: token.ASSIGN, expectedLiteral: "="},
		{expectedType: token.INT, expectedLiteral: "5"},
		{expectedType: token.CONST, expectedLiteral: "insist"},
		{expectedType: token.IDENT, expectedLiteral: "ten"},
		{expectedType: token.ASSIGN, expectedLiteral: "="},
		{expectedType: token.INT, expectedLiteral: "10"},
		{expectedType: token.CONST, expectedLiteral: "insist"},
		{expectedType: token.IDENT, expectedLiteral: "result"},
		{expectedType: token.ASSIGN, expectedLiteral: "="},
		{expectedType: token.IDENT, expectedLiteral: "summing"},
		{expectedType: token.LPAREN, expectedLiteral: "("},
		{expectedType: token.IDENT, expectedLiteral: "five"},
		{expectedType: token.COMMA, expectedLiteral: ","},
		{expectedType: token.IDENT, expectedLiteral: "ten"},
		{expectedType: token.RPAREN, expectedLiteral: ")"},
		{expectedType: token.EOF, expectedLiteral: ""},
	}

	l := New(input)

	for i, tt := range tests {
		tok := l.NextToken()

		if tok.Type != tt.expectedType {
			t.Fatalf("tests [%d] - token type wrong. expected=%q, got=%q", i, tt.expectedType, tok.Type)
		}

		if tok.Literal != tt.expectedLiteral {
			t.Fatalf("tests[%d] - literal wrong. expected=%q, got=%q", i, tt.expectedLiteral, tok.Literal)
		}
	}
}
